crates/proto/src/rr/rdata/cert.rs

```diff
@@ -0,0 +1,708 @@
+            Algorithm::Unassigned(value) if (18..=22).contains(&value) => value,
+            Algorithm::ECCGOST12 => 23,
+            Algorithm::Unassigned(value) if (24..=122).contains(&value) => value,
+            Algorithm::INDIRECT => 252,
+            Algorithm::PRIVATEDNS => 253,
+            Algorithm::PRIVATEOID => 254,
+            Algorithm::Unassigned(value) => value,
+            Algorithm::Reserved(value) => value,
+        }
+    }
+}
+
+impl fmt::Display for Algorithm {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+/// [RFC 4398, Storing Certificates in DNS, November 1987][rfc4398]
+/// https://tools.ietf.org/html/rfc4398
+///
+/// ```text
+///
+/// The CERT resource record (RR) has the structure given below.  Its RR
+/// type code is 37.
+///
+///    1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+/// 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |             type              |             key tag           |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |   algorithm   |                                               /
+/// +---------------+            certificate or CRL                 /
+/// /                                                               /
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-|
+/// ```
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct CERT {
+    cert_type: CertType,
+    key_tag: u16,
+    algorithm: Algorithm,
+    cert_data: Vec<u8>,
+}
+
+impl CERT {
+    /// Construct a new CERT RData
+    pub const fn new(
+        cert_type: CertType,
+        key_tag: u16,
+        algorithm: Algorithm,
+        cert_data: Vec<u8>,
+    ) -> Self {
+        Self {
+            cert_type,
+            key_tag,
+            algorithm,
+            cert_data,
+        }
+    }
+
+    /// Returns the CERT type
+    pub fn cert_type(&self) -> CertType {
+        self.cert_type
+    }
+
+    /// Returns the CERT key tag
+    pub fn key_tag(&self) -> u16 {
+        self.key_tag
+    }
+
+    /// Returns the CERT algorithm
+    pub fn algorithm(&self) -> Algorithm {
+        self.algorithm
+    }
+
+    /// Returns the CERT record data
+    pub fn cert_data(&self) -> Vec<u8> {
+        self.cert_data.clone()
+    }
+
+    /// Returns the CERT (Base64)
+    pub fn cert_base64(&self) -> String {
+        data_encoding::BASE64.encode(&self.cert_data).clone()
+    }
+}
+
+impl TryFrom<&[u8]> for CERT {
+    type Error = ProtoError;
+
+    fn try_from(cert_record: &[u8]) -> Result<Self, Self::Error> {
+        if cert_record.len() <= 5 {
+            return Err(ProtoError::from("invalid cert_record length".to_string()));
+        }
+
+        let cert_type = ((cert_record[0] as u16) << 8) | (cert_record[1] as u16);
```

Comment:
This reads are a little redundant to what we have out of the box in the library for reading data.

For example, you can use the `decoder.read_u16()` in the RecordDataDecodable. Similar with the next u16. The following one with the algorithm looks like a u32, `decoder.read_u32`.

It helps guarantee bounds and such. Since you need the unknown length of the record, what I've done in other record types like this, is track the index of the current data being read with `let start_idx = decoder.index();` then collect all the other interior data with the library methods, and finally, 

```
let len = length
        .map(|u| u as usize)
        .checked_sub(decoder.index() - start_idx)
        .map_err(|_| ProtoError::from("invalid rdata length in SIG"))?
        .unverified(/*used only as length safely*/);

let cert = decoder
        .read_vec(sig_len)?
        .unverified(/*will fail in usage if invalid*/);
```

Reply:
You can see an example here: https://github.com/hickory-dns/hickory-dns/blob/main/crates/proto/src/rr/dnssec/rdata/sig.rs#L498-L532

On top of that, you'll notice that Algorithm can be read with it's internal method (I think it's the same type):

```
let algorithm = Algorithm::read(decoder)?;
```

Reply:
btw, I think your logic here is technically correct, but we've had a number of dirty data issues that a lot of the vigilance in the supporting library routines helps eliminate.

Reply:
I refactored the read_data to use the decoder and refactored try_from to use read_data